07-MemoryManagementKernelModule/allocation.c

```diff
@@ -0,0 +1,275 @@
+			deltaFreeNS[i] = deltaFree[i].tv_nsec;
+			if (deltaFree[i].tv_sec != 0)
+				deltaFreeNS[i] += 1000000000;
+		}
+
+		bubbleSort(deltaAllocNS, attempts);
+		bubbleSort(deltaFreeNS, attempts);
+
+		pr_info("kmalloc:2^%02d+%d	%09ld	%09ld	%09ld	%09ld	%09ld	%09ld",
+			power, oversize,
+			deltaAllocNS[1], deltaAllocNS[attempts/2],
+			deltaAllocNS[attempts - 2],
+			deltaFreeNS[1], deltaFreeNS[attempts/2],
+			deltaFreeNS[attempts - 2]);
+
+		oversize++;
+		oversize %= 2;
+		if (oversize == 0)
+			power++;
+	}
+
+	power = 0;
+	oversize = 0;
+
+	pr_info("kzalloc:size	minAlloc	avrAlloc	maxAlloc	minFree		avrFree		maxFree");
+
+	while (power < 65) {
+		const int attempts = 9;
+		struct timespec t1, t2;
+		struct timespec deltaAlloc[attempts], deltaFree[attempts];
+		unsigned long long size;
+
+		int i = 0;
+
+		for (; i < attempts; i++) {
+			size = (1 << power) + oversize;
+			getnstimeofday(&t1);
+			ptr = kzalloc(size, GFP_KERNEL);
+			getnstimeofday(&t2);
+			deltaAlloc[i] = sub(t2, t1);
+
+			getnstimeofday(&t1);
+			kfree(ptr);
+			getnstimeofday(&t2);
+			deltaFree[i] = sub(t2, t1);
+		}
+
+		long deltaAllocNS[attempts], deltaFreeNS[attempts];
+
+		i = 0;
+		for (; i < attempts; i++) {
+			deltaAllocNS[i] = deltaAlloc[i].tv_nsec;
+			if (deltaAlloc[i].tv_sec != 0)
+				deltaAllocNS[i] += 1000000000;
+			deltaFreeNS[i] = deltaFree[i].tv_nsec;
+			if (deltaFree[i].tv_sec != 0)
+				deltaFreeNS[i] += 1000000000;
+		}
+
+		bubbleSort(deltaAllocNS, attempts);
+		bubbleSort(deltaFreeNS, attempts);
+
+		pr_info("kzalloc:2^%02d+%d	%09ld	%09ld	%09ld	%09ld	%09ld	%09ld",
+			power, oversize,
+			deltaAllocNS[1], deltaAllocNS[attempts/2],
+			deltaAllocNS[attempts - 2],
+			deltaFreeNS[1], deltaFreeNS[attempts/2],
+			deltaFreeNS[attempts - 2]);
+
+		oversize++;
+		oversize %= 2;
+		if (oversize == 0)
+			power++;
+	}
+
+	power = 0;
+	oversize = 0;
+
+	pr_info("vmalloc:size	minAlloc	avrAlloc	maxAlloc	minFree		avrFree		maxFree");
+	while (power < 65) {
+		const int attempts = 9;
+		struct timespec t1, t2;
+		struct timespec  deltaAlloc[attempts], deltaFree[attempts];
+		unsigned long long size;
+
+		int i = 0;
+
+		for (; i < attempts; i++) {
+			size = (1 << power) + oversize;
+			getnstimeofday(&t1);
+			ptr = vmalloc(size);
+			getnstimeofday(&t2);
+			deltaAlloc[i] = sub(t2, t1);
+
+			getnstimeofday(&t1);
+			vfree(ptr);
+			getnstimeofday(&t2);
+			deltaFree[i] = sub(t2, t1);
+		}
+
+		long deltaAllocNS[attempts], deltaFreeNS[attempts];
+
+		i = 0;
+		for (; i < attempts; i++) {
+			deltaAllocNS[i] = deltaAlloc[i].tv_nsec;
+			if (deltaAlloc[i].tv_sec != 0)
+				deltaAllocNS[i] += 1000000000;
+			deltaFreeNS[i] = deltaFree[i].tv_nsec;
+			if (deltaFree[i].tv_sec != 0)
+				deltaFreeNS[i] += 1000000000;
+		}
+
+		bubbleSort(deltaAllocNS, attempts);
+		bubbleSort(deltaFreeNS, attempts);
+
+		pr_info("vmalloc:2^%02d+%d	%09ld	%09ld	%09ld	%09ld	%09ld	%09ld",
+			power, oversize,
+			deltaAllocNS[1], deltaAllocNS[attempts/2],
```

Comment:
Please consider the difference between average and median value. Is it require to sort the obtained results?